imap_processing/cdf/cdf_attribute_manager.py

```diff
@@ -0,0 +1,192 @@
+"""Class for creating and managing CDF attrs. Developed based of HermesDataSchema from HERMES-SOC/hermes_core."""
+from __future__ import annotations
+
+from collections import OrderedDict
+from pathlib import Path
+import yaml
+from imap_processing.cdf import const
+import numpy as np
+
+DEFAULT_GLOBAL_CDF_ATTRS_FILE = "imap_default_global_cdf_attrs.yaml"
+DEFAULT_GLOBAL_CDF_ATTRS_SCHEMA_FILE = "hermes_default_global_cdf_attrs_schema.yaml"
+DEFAULT_VARIABLE_CDF_ATTRS_SCHEMA_FILE = "hermes_default_variable_cdf_attrs_schema.yaml"
+
+class CdfAttributeManager:
+    def __init__(self, data_dir: Path):
+
+        # TODO: Split up schema source and data source?
+        self.source_dir = data_dir
+
+        # TODO: copied from hermes_core. Currently we can use default schema, but
+        # We should add some way of extending the schema and remove all the HERMES specific stuff
+        # Data Validation, Complaiance, Derived Attributes
+        self.global_attribute_schema = self._load_default_global_attr_schema()
+
+        # Data Validation and Compliance for Variable Data
+        self.variable_attribute_schema = (
+            self._load_default_variable_attr_schema()
+        )
+
+        self.cdftypenames = {
```

Comment:
This is interesting, because it is using ctypes, which is specific to `spacepy`'s cdf writer and not cdflib which we are using... So I'm curious if we need this or what the value is for the IMAP project rather than numpy types, or are we going to run into any other issues?

Reply:
This is from the HERMES soc so I'm not sure! I can remove this part and leave the numpy types...? I'm not really sure what IMAP will end up needing vs what we would want in the generic version of the code.

Reply:
I didn't follow what this was actually used for. My assumption is some kind of validation, which makes me think that it might help HERMES validation because they use `spacepy` and these data types, but not us because we use a different set of datatypes. So, I think this is up to you whether this is needed or not, it just stuck out to me as something that might be more project-specific.

Reply:
We use this for metadata derivations leveraging some of the `spacepy` code! There is some code that will help us identify the CDF variable types based on the numpy types. We then use this for setting default values for VALIDMIN, VALIDMAX, and FORMAT attributes. You shouldn't need this or the `const.py` if you're not going to use this for any attribute derivations using `spacepy`.